File: src/main/java/at/searles/parsing/Ref.java
package at.searles.parsing;

import at.searles.parsing.printing.ConcreteSyntaxTree;

public class Ref<T> implements Parser<T> {

    private final String label;
    private Parser<T> ref;

    public Ref(String label) {
        this.label = label;
    }

    public Ref<T> set(Parser<T> ref) {
        this.ref = ref;
        return this;
    }

    @Override
    public boolean recognize(ParserStream stream) {
        try {
            return ref.recognize(stream);
        } catch(StackOverflowError e) {
            throw new PossiblyInfiniteRecursionException(this, e);
        }
    }

    @Override
    public T parse(ParserStream stream) {
        try {
            return ref.parse(stream);
        } catch(StackOverflowError e) {
            throw new PossiblyInfiniteRecursionException(this, e);
        }
    }

    @Override
    public ConcreteSyntaxTree print(T t) {
        try {
            return ref.print(t);
        } catch(StackOverflowError e) {
            throw new PossiblyInfiniteRecursionException(this, e);
        }
    }

    @Override
    public String toString() {
        return label;
    }
}

File: src/main/java/at/searles/parsing/Ref.kt
package at.searles.parsing

import at.searles.parsing.printing.ConcreteSyntaxTree

class Ref<T>(private val label: String) : Parser<T> {

    lateinit var ref: Parser<T>

    override fun recognize(stream: ParserStream): Boolean {
        return try {
            ref.recognize(stream)
        } catch (e: StackOverflowError) {
            throw PossiblyInfiniteRecursionException(this, e)
        }
    }

    override fun parse(stream: ParserStream): T? {
        return try {
            ref.parse(stream)
        } catch (e: StackOverflowError) {
            throw PossiblyInfiniteRecursionException(this, e)
        }
    }

    override fun print(item: T): ConcreteSyntaxTree? {
        return try {
            ref.print(item)
        } catch (e: StackOverflowError) {
            throw PossiblyInfiniteRecursionException(this, e)
        }
    }

    override fun toString(): String {
        return label
    }

}